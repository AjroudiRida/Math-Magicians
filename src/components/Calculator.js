import { useState, useEffect } from 'react';
import calculate from '../logic/calculate';
import CalculatorBtns from './CalculatorBtns';
import Quotes from './Quotes';
import './calculator.css';

const Calculator = () => {
  const [obj, setObj] = useState({
    total: null,
    next: null,
    operation: null,
  });
  const [inputValue, setInputValue] = useState('0');

  const getBtnValue = (btnText) => {
    const objState = calculate(obj, btnText);
    setObj(objState);
  };

  useEffect(() => {
    if (obj.next) {
      setInputValue(obj.next);
    }

    if (obj.total && obj.next) {
      setInputValue(obj.next);
    }

    if (obj.total && !obj.next) {
      setInputValue(obj.total);
    }

    if (!obj.next && !obj.total) {
      setInputValue('0');
    }
  }, [obj]);

  return (
    <div className="main-section container">
      <div className="section-quotes">
        <Quotes />
      </div>
      <div className="section-calculator">
        <div className="calculator">
          <div className="input" contentEditable={false}>
            {inputValue}
          </div>
          <div className="calculator-functions">
            <CalculatorBtns
              content="AC"
              handleClick={() => {
                getBtnValue('AC');
              }}
            />
            <CalculatorBtns
              content="+/-"
              handleClick={() => {
                getBtnValue('+/-');
              }}
            />
            <CalculatorBtns
              content="%"
              handleClick={() => {
                getBtnValue('%');
              }}
            />
            <CalculatorBtns
              content="รท"
              handleClick={() => {
                getBtnValue('รท');
              }}
            />
            <CalculatorBtns
              content="7"
              handleClick={() => {
                getBtnValue('7');
              }}
            />
            <CalculatorBtns
              content="8"
              handleClick={() => {
                getBtnValue('8');
              }}
            />
            <CalculatorBtns
              content="9"
              handleClick={() => {
                getBtnValue('9');
              }}
            />
            <CalculatorBtns
              content="x"
              handleClick={() => {
                getBtnValue('x');
              }}
            />
            <CalculatorBtns
              content="4"
              handleClick={() => {
                getBtnValue('4');
              }}
            />
            <CalculatorBtns
              content="5"
              handleClick={() => {
                getBtnValue('5');
              }}
            />
            <CalculatorBtns
              content="6"
              handleClick={() => {
                getBtnValue('6');
              }}
            />
            <CalculatorBtns
              content="-"
              handleClick={() => {
                getBtnValue('-');
              }}
            />
            <CalculatorBtns
              content="1"
              handleClick={() => {
                getBtnValue('1');
              }}
            />
            <CalculatorBtns
              content="2"
              handleClick={() => {
                getBtnValue('2');
              }}
            />
            <CalculatorBtns
              content="3"
              handleClick={() => {
                getBtnValue('3');
              }}
            />
            <CalculatorBtns
              content="+"
              handleClick={() => {
                getBtnValue('+');
              }}
            />
            <CalculatorBtns
              content="0"
              handleClick={() => {
                getBtnValue('0');
              }}
            />
            <CalculatorBtns
              content="."
              handleClick={() => {
                getBtnValue('.');
              }}
            />
            <CalculatorBtns
              content="="
              handleClick={() => {
                getBtnValue('=');
              }}
            />
          </div>
        </div>
      </div>
    </div>
  );
};

export default Calculator;
