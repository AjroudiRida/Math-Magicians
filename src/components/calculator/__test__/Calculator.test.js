import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import Calculator from '../Calculator';

describe('calculator tests', () => {
  it('should match the Dom snapshot', () => {
    const calculatorTree = renderer.create(<Calculator />).toJSON();
    expect(calculatorTree).toMatchSnapshot();
  });

  it('should render in the document', () => {
    render(<Calculator />);
    const calculator = screen.getByTestId('calc');
    expect(calculator).toBeInTheDocument();
  });

  it('changes the input value when a button is clicked', () => {
    render(<Calculator />);
    const btnElement = screen.getAllByTestId('button').filter((button) => button.textContent === '5');
    fireEvent.click(btnElement[0]);
    const inputElement = screen.getByTestId('input', { name: '5' });
    expect(inputElement.textContent).toBe('5');
  });
});
