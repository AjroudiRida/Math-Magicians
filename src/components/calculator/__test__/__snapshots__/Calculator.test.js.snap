// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculator tests should match the Dom snapshot 1`] = `
<div
  className="main main-section"
  data-testid="calc"
>
  <div
    className="section-quotes"
  >
    <h1>
      Let's do some math!
    </h1>
  </div>
  <div
    className="section-calculator"
  >
    <div
      className="calculator"
    >
      <div
        className="input"
        contentEditable={false}
        data-testid="input"
      >
        0
      </div>
      <div
        className="calculator-functions"
      >
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            AC
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            +/-
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            %
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            รท
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            7
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            8
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            9
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            x
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            4
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            5
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            6
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            -
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            1
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            2
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            3
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            +
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            0
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            .
          </span>
        </button>
        <button
          className="btn"
          data-testid="button"
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <span>
            =
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
